### ISSUE TITLE
DaD user designer

### ISSUE THREAD

**Issue by pofider** [jsreport/jsreport#121]
Implement client side component allowing to the non-developers to design simple reports. This component should be pluggable to any kind of frontend and produce jsreport compatible templates.

**Comment by pofider** [comment]
The goal is to provide an easy to use end user report visual designer. Such a designer should be a standalone component producing jsreport compatible templates on the background. It is expected this component will be heavily integrated into the external applications thus it should be designed to support custom branding and extensions. Technically this should be the completely separate component with the brand new web stack using React and ES6.

The designer should produce design meta definition and along the real handlebars based jsreport template. It should always transform meta definition into the template every time the user changes the design. 

The designer ui will consists of an designer pane, command bar and various widgets for designing tables, charts, images...

**Designer Pane**
We should probably use grid system for the report rather than pixel based absolute positioning. User should be able to drag and drop or re-size the widgets on the pane.

**Widgets**
Every widget should provide the following:
- static html (icon) used on the command bar
- html used on the design pane
- handlebars template used for the real rendering
- ui for filling the settings

**Data Source**
The integrator of the component should provide json based schema of the data source to the designer. The designer then uses the schema to build a visual tree component which will be used in widgets to define the data binding.

**Basic data sources**
The designer should later provide some direct data sources like sql, nosql...

**Previewing**
There will be several ways to invoke preview of the report and it is up to the integrator which one chooses.
- preview based directly on the schema json
- integrator completely takes the responsibility of previewing
- integrator provides dynamic input data
- integrator provides custom script loading data

**Advanced editing**
User should not be limited by the provided widgets and its capabilities. He should be able to alter directly the handlebars produced by the widget. In this case the designer warns the user that this widget will be transformed with no way back into the different widget type supporting html edits. 
 
So this was just to provide the basic requirements and concepts. We should start with prototyping a designer pane with support for dad a resizing widgets.

**Comment by pofider** [comment]
This could be an inspiration
https://github.com/rsamec/react-designer

**Comment by bjrmatos** [comment]
great notes!

> Designer Pane
> We should probably use grid system for the report rather than pixel based absolute positioning. User should be able to drag and drop or re-size the widgets on the pane.

totally agree

another inspiration
https://github.com/ipselon/structor

**Comment by pofider** [comment]
So I also found https://slides.com/ which is very nicely solving very similar problem, unfortunately not open sourced. After long searching I didn't find a reusable component that would fit our needs for drag and drop designer. It looks we will need to do this from the scratch.

I've also spend some time with extending my reactjs knowledge and also trying some prototyping with [react-dnd](https://github.com/gaearon/react-dnd) however I still don't feel like and expert for such a react task. I don't want to screw its architecture just at the very start. This is probably just too big to learn react on it.

In other words, @bjrmatos , don't you want this to be your main task, along with other smaller ones? I believe you are way far in the react then I am and the best one for this.

**Comment by bjrmatos** [comment]
yes! i can do this, just let me finish `electron-html-to-pdf`, i really want to close the idea of using electron as a rendering platform (god i wish i have more time) 

can you create a repository under jsreport org? or would you like me to do this in a repository under my account?

**Comment by pofider** [comment]
Great. Take your time with electron-html-to-pdf. This will be anyway long term task, so there is no hurry.

You can use repository [jsreport-designer](https://github.com/jsreport/jsreport-designer).

**Comment by bjrmatos** [comment]
This video has inspired me, That UI designer looks very solid! https://youtu.be/PggnIYdaJFI

**Comment by pofider** [comment]
Just linking here that we already fully work on the designer feature in the mentioned repository https://github.com/jsreport/jsreport-designer

**Comment by deceiver08** [comment]
Hi All. Just checking in on this. I totally want to implement jsreports, but my end users need to make their own reports through a wysiwyg type interface. I think the jsreport community has been waiting patiently for this for a **long** time and would be ecstatic to see this feature come to reality.

**Comment by fmquaglia** [comment]
Hey hello. Is there an update on the Drag and Drop feature? 
It's the only thing is stopping us to use this.
